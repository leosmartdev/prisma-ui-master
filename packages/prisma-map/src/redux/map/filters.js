import { getType } from '@turf/turf';

/*
 * COPYRIGHT Â© 2018 OROLIA GROUP AND/OR ITS AFFILIATES. ALL RIGHTS ARE STRICTLY RESERVED.
 * THIS SOURCE CODE CONTAINS PROPRIETARY AND CONFIDENTIAL INFORMATION AND DATA AND IS THE
 * SOLE PROPERTY OF OROLIA GROUP AND/OR ITS AFFILIATES.
 * THIS SOURCE CODE MUST NOT BE USED, DISSEMINATED, OR DISTRIBUTED EXCEPT FOR THE AGREED PURPOSE.
 * UNAUTHORIZED USE, REPRODUCTION, OR ISSUE TO ANY THIRD PARTY IS NOT PERMITTED WITHOUT THE
 * PRIOR WRITTEN CONSENT OF THE OROLIA GROUP.
 * THIS SOURCE CODE IS TO BE RETURNED TO THE OROLIA GROUP WHEN THE AGREED PURPOSE IS FULFILLED.
 * ------------------------------------
 *
 * Collection of pure functions for filtering features into a layer. These are the implementations
 * of the `state['@prisma/map'].layers[id].filters` filters.
 */

/**
 * Runs the feature against the filter provided on the layer and returns the result.
 * If the filter is an empty array, that will always match and return true. The array is the filter
 * funtion followed by parameters for that function. To run multiple filters, use the `all` or `any`
 * filters to run && or || comparisons.
 *
 * @param {object} feature the feature to check against the filters.
 * @param {array} filters List containing the filter function and params to run.
 * @return {boolean} The boolean result of the filter run.
 */
export function runLayerFilter(feature, filter) {
  // Empty filter matches everything
  if (filter.length < 1) {
    return true;
  }

  // Get filter function, it should always be the first thing.
  const [filterFunc, ...params] = filter;

  // return runFilterFunction(feature, filterFunc, ...params);
  switch (filterFunc) {
    case 'all': return allFilter(feature, ...params);
    case 'any': return anyFilter(feature, ...params);
    case 'featureProperty': return featureProperty(feature, ...params);
    case 'featureType': return featureType(feature, ...params);
    default: return false;
  }
}

/**
 * AND boolean operation between all the filters passed in. If any of the filters passed are false,
 * the function returns false.
 *
 * @param {object} feature The feature
 * @param {array(filters)} filters The additional filters to be and'ed together.
 * @return {boolean} true if all sub filters are true, false if any of the sub filters are false.
 */
function allFilter(feature, ...filters) {
  return filters.every(filter => runLayerFilter(feature, filter));
}

/**
 * OR boolean operation between all the filters passed in. If any of the filters passed are true,
 * the function returns true, if all are false, then returns false.
 *
 * @param {object} feature The feature
 * @param {array(filters)} filters The additional filters to be or'ed together.
 * @return {boolean} true if any sub filters are true, false if all of the sub filters are false.
 */
function anyFilter(feature, ...filters) {
  return filters.some(filter => runLayerFilter(feature, filter));
}

/**
 *
 * Eg:
 *   `feature.properties[property] == value`
 *   `feature.properties[property] < value`
 *
 * Note: == will use the underlying === for the comparison.
 *
 * @param {object} feature The feature the filter is being applied to.
 * @param {string} operator boolean operator, oneof '==', '!=', '<=', '>=', '<', '>'
 * @param {string} property The property name to retrieve from `feature.properties`
 * @param {any} value The value to compare the retrieved property value against.
 */
export function featureProperty(feature, operator, property, value) {
  if (!feature.properties) {
    return false;
  }

  const featureValue = feature.properties[property];
  switch (operator) {
    case '==': return featureValue === value;
    case '!=': return featureValue !== value;
    case '<': return featureValue < value;
    case '>': return featureValue > value;
    case '<=': return featureValue <= value;
    case '>=': return featureValue >= value;
    default:
      return false;
  }
}

/**
 * Takes the provided feature and returns if it is of the provided type. If invalid feature is
 * passed in, will return false.
 *
 * Eg:
 * ```
 * feature.geometry.type == 'Point'
 * featureType(feature, 'Point') === true
 * featureType(feature, 'Polygon') === false
 * ```
 *
 * @param {object} feature The feature the filter is being applied to.
 * @param {string} type GeoJSON type to compare the feature to.
 * @return {boolean} true if the feature type matches the provided type.
 */
export function featureType(feature, type) {
  try {
    return getType(feature) === type;
  } catch (e) {
    return false;
  }
}
