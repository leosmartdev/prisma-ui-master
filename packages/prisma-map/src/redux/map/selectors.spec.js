import { point } from '@turf/turf';
import deepFreeze from 'deep-freeze';
import {
  isFeatureOnLayer,
  createStylesForAllMaps,
  getGeneratedStyleForMap,
  getStylesForLoadedMaps,
  getLayerConfig,
  getFlyToPositionForMap,
  getModeForMap,
} from './selectors';
import { runLayerFilter } from './filters';

jest.mock('./filters', () => ({
  runLayerFilter: jest.fn(() => false),
}));

describe('redux/map/selectors', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('createStylesForAllMaps', () => {
    const state = {
      map: {
        mapIds: ['foo'],
        foo: {
          loaded: true,
          style: {
            mapId: 'foo',
            sources: {
              url: {
                type: 'geojson',
                url: 'http://foo.com',
              },
            },
            layers: [
              { id: 'layer1', source: 'url' },
            ],
          },
          layers: {
            dynamicLayers: ['layer2'],
          },
        },
      },
      layers: {
        layer2: {
          id: 'layer2',
          name: 'Layer 2',
          source: 'features',
        },
      },
    };

    it('adds features to proper layer sources', () => {
      deepFreeze(state);

      runLayerFilter.mockImplementation(() => true);
      const f1 = point([0, 1]);
      f1.properties.bar = 'baz';
      const f2 = point([1, 2]);
      f2.properties.bar = 'foo';

      const testState = {
        '@prisma/map': {
          ...state,
          layers: {
            ...state.layers,
            layer2: {
              ...state.layers.layer2,
              filter: ['featureProperty', '!=', 'bar', undefined],
            },
          },
          features: {
            f1,
            f2,
          },
        },
      };

      deepFreeze(testState);

      const generatedStyles = createStylesForAllMaps(testState);

      expect(generatedStyles).toMatchSnapshot();
    });

    it('only adds a feature to the first layer that matches', () => {
      runLayerFilter.mockImplementation((feature, filter) => filter[0]);
      const f1 = point([0, 1]);
      f1.properties.layerId = 'layer3';

      const testState = {
        '@prisma/map': {
          ...state,
          map: {
            ...state.map,
            foo: {
              ...state.map.foo,
              layers: {
                dynamicLayers: ['layer2', 'layer3', 'layer4'],
              },
              style: {
                ...state.map.foo.style,
                layers: [
                  { id: 'layer1', source: 'url' },
                ],
              },
            },
          },
          layers: {
            layer2: {
              source: 'features',
              filter: [false],
            },
            layer3: {
              source: 'features',
              filter: [true],
            },
            layer4: {
              source: 'features',
              filter: [true],
            },
          },
          features: {
            f1,
          },
        },
      };

      deepFreeze(testState);

      const generatedStyles = createStylesForAllMaps(testState);

      // expecting layer 4 to match in this case.
      expect(generatedStyles).toMatchSnapshot();
    });

    it('add features to all maps that match', () => {
      runLayerFilter.mockImplementation((feature, filter) => filter[0]);
      const f1 = point([0, 1]);
      f1.properties.layerId = 'layer3';

      const map = {
        ...state.map.foo,
        layers: {
          dynamicLayers: ['layer2', 'layer3', 'layer4', 'layer5'],
        },
        style: {
          ...state.map.foo.style,
          layers: [
            { id: 'layer1', source: 'url' },
          ],
        },
      };

      const testState = {
        '@prisma/map': {
          map: {
            mapIds: ['foo', 'bar', 'baz'],
            foo: { ...map },
            bar: {
              ...map,
              layers: {
                dynamicLayers: ['layer2', 'layer5'],
              },
              style: {
                ...state.map.foo.style,
                mapId: 'bar',
                layers: [],
              },
            },
            baz: {
              ...map,
              layers: {
                dynamicLayers: ['layer2', 'layer4'],
              },
              style: {
                ...state.map.foo.style,
                mapId: 'baz',
                layers: [],
              },
            },
          },
          layers: {
            layer2: {
              source: 'features',
              filter: [false],
            },
            layer3: {
              source: 'features',
              filter: [true],
            },
            layer4: {
              source: 'features',
              filter: [false],
            },
            layer5: {
              source: 'features',
              filter: [true],
            },
          },
          features: {
            f1,
          },
        },
      };

      deepFreeze(testState);

      const generatedStyles = createStylesForAllMaps(testState);

      expect(generatedStyles).toMatchSnapshot();
    });
  });

  describe('getStylesForLoadedMaps()', () => {
    it('only returns generated styles for loaded maps', () => {
      const testState = {
        '@prisma/map': {
          map: {
            mapIds: ['foo', 'bar', 'baz'],
            foo: {
              loaded: true,
              style: {
                mapId: 'foo',
                sources: {},
                layers: [],
              },
            },
            bar: {
              loaded: true,
              style: {
                mapId: 'bar',
                sources: {},
                layers: [],
              },
            },
            baz: {
              loaded: false,
              style: {
                mapId: 'bar',
                sources: {},
                layers: [],
              },
            },
          },
        },
      };

      deepFreeze(testState);

      const generatedStyles = getStylesForLoadedMaps(testState);

      expect(generatedStyles).toMatchSnapshot();
    });

    it('adds all sources and layers listed in dynamicLayers', () => {
      const testState = {
        '@prisma/map': {
          map: {
            mapIds: ['foo'],
            foo: {
              loaded: true,
              style: {
                mapId: 'foo',
                sources: {},
                layers: [],
              },
              layers: {
                dynamicLayers: ['layer1', 'layer2'],
                layer1: {
                  type: 'circle',
                  name: 'Layer 1',
                  source: 'features',
                },
                layer2: {
                  type: 'fill',
                  name: 'Layer 2',
                  source: 'features',
                },
              },
            },
          },
          layers: {},
        },
      };

      deepFreeze(testState);

      const generatedStyles = getStylesForLoadedMaps(testState);

      expect(generatedStyles).toMatchSnapshot();
    });

    it('adds dynamic layers after the existing ones', () => {
      const testState = {
        '@prisma/map': {
          map: {
            mapIds: ['foo'],
            foo: {
              loaded: true,
              style: {
                mapId: 'foo',
                sources: {},
                layers: [{ id: 'existing' }],
              },
              layers: {
                dynamicLayers: ['layer1', 'layer2'],
                layer1: {
                  type: 'circle',
                  name: 'Layer 1',
                  source: 'features',
                },
                layer2: {
                  type: 'fill',
                  name: 'Layer 2',
                  source: 'features',
                },
              },
            },
          },
          layers: {},
        },
      };

      deepFreeze(testState);

      const generatedStyles = getStylesForLoadedMaps(testState);

      expect(generatedStyles).toEqual(expect.objectContaining({
        foo: {
          mapId: 'foo',
          layers: [
            { id: 'existing' },
            {
              id: 'layer1',
              type: 'circle',
              source: 'generated:layer1',
            }, {
              id: 'layer2',
              type: 'fill',
              source: 'generated:layer2',
            },
          ],
          sources: expect.any(Object),
        },
      }));
    });

    it('doesnt add layers not listed in dynamicLayers ', () => {
      const testState = {
        '@prisma/map': {
          map: {
            mapIds: ['foo'],
            foo: {
              loaded: true,
              style: {
                mapId: 'foo',
                sources: {},
                layers: [],
              },
              layers: {
                dynamicLayers: ['layer2'],
                layer1: {
                  type: 'circle',
                  name: 'Layer 1',
                  source: 'features',
                },
                layer2: {
                  type: 'fill',
                  name: 'Layer 2',
                  source: 'features',
                  paint: {
                    'circle-color': '#FFFFFF',
                  },
                  layout: {
                    visibility: 'none',
                  },
                },
              },
            },
          },
          layers: {},
        },
      };

      deepFreeze(testState);

      const generatedStyles = getStylesForLoadedMaps(testState);

      expect(generatedStyles).toEqual(expect.objectContaining({
        foo: {
          mapId: 'foo',
          layers: [{
            id: 'layer2',
            type: 'fill',
            source: 'generated:layer2',
            paint: {
              'circle-color': '#FFFFFF',
            },
            layout: {
              visibility: 'none',
            },
          }],
          sources: expect.any(Object),
        },
      }));
    });
  });

  describe('getLayerConfig()', () => {
    it('returns global layer config', () => {
      const state = {
        '@prisma/map': {
          map: {
            bar: {
              layers: {},
            },
          },
          layers: {
            foo: {
              id: 'foo',
              source: 'historicalFeatures',
              type: 'circle',
              filter: ['featureType', 'Point'],
              layout: { visibility: 'visible' },
              paint: { 'circle-color': '#00FF00' },
            },
          },
        },
      };
      deepFreeze(state);

      const layerConfig = getLayerConfig(state, 'bar', 'foo');

      expect(layerConfig).toEqual({
        id: 'foo',
        source: 'historicalFeatures',
        type: 'circle',
        filter: ['featureType', 'Point'],
        layout: { visibility: 'visible' },
        paint: { 'circle-color': '#00FF00' },
      });
    });

    it('returns map layer config', () => {
      const state = {
        '@prisma/map': {
          map: {
            bar: {
              layers: {
                foo: {
                  id: 'foo',
                  source: 'historicalFeatures',
                  type: 'circle',
                  filter: ['featureType', 'Point'],
                  layout: { visibility: 'visible' },
                  paint: { 'circle-color': '#00FF00' },
                },
              },
            },
          },
          layers: {},
        },
      };
      deepFreeze(state);

      const layerConfig = getLayerConfig(state, 'bar', 'foo');

      expect(layerConfig).toEqual({
        id: 'foo',
        source: 'historicalFeatures',
        type: 'circle',
        filter: ['featureType', 'Point'],
        layout: { visibility: 'visible' },
        paint: { 'circle-color': '#00FF00' },
      });
    });

    it('returns style ready config when passed styleLayer=true', () => {
      const state = {
        '@prisma/map': {
          map: {
            bar: {
              layers: {
                foo: {
                  id: 'foo',
                  source: 'historicalFeatures',
                  type: 'circle',
                  filter: ['featureType', 'Point'],
                  layout: { visibility: 'visible' },
                  paint: { 'circle-color': '#00FF00' },
                  other: 'properties',
                },
              },
            },
          },
          layers: {},
        },
      };
      deepFreeze(state);

      const layerConfig = getLayerConfig(state, 'bar', 'foo', true);

      expect(layerConfig).toEqual({
        id: 'foo',
        source: 'generated:foo',
        type: 'circle',
        layout: { visibility: 'visible' },
        paint: { 'circle-color': '#00FF00' },
      });
    });

    it('merges global with map config', () => {
      const state = {
        '@prisma/map': {
          map: {
            bar: {
              layers: {
                foo: {
                  id: 'foo',
                  layout: { visibility: 'visible' },
                },
              },
            },
          },
          layers: {
            foo: {
              id: 'foo',
              type: 'circle',
              filter: ['featureType', 'Point'],
              paint: { 'circle-color': '#00FF00' },
            },
          },
        },
      };
      deepFreeze(state);

      const layerConfig = getLayerConfig(state, 'bar', 'foo', true);

      expect(layerConfig).toEqual({
        id: 'foo',
        source: 'generated:foo',
        type: 'circle',
        layout: { visibility: 'visible' },
        paint: { 'circle-color': '#00FF00' },
      });
    });

    it('returns config with map config overriding global config styleLayer = true', () => {
      const state = {
        '@prisma/map': {
          map: {
            bar: {
              layers: {
                foo: {
                  id: 'foo',
                  paint: { 'circle-color': '#000000' },
                  layout: { visibility: 'visible' },
                },
              },
            },
          },
          layers: {
            foo: {
              id: 'foo',
              type: 'circle',
              filter: ['featureType', 'Point'],
              paint: { 'circle-color': '#00FF00' },
              layout: { visibility: 'none' },
            },
          },
        },
      };
      deepFreeze(state);

      const layerConfig = getLayerConfig(state, 'bar', 'foo', true);

      expect(layerConfig).toEqual({
        id: 'foo',
        source: 'generated:foo',
        type: 'circle',
        layout: { visibility: 'visible' },
        paint: { 'circle-color': '#000000' },
      });
    });

    it('returns config with map config overriding global config styleLayer = false', () => {
      const state = {
        '@prisma/map': {
          map: {
            bar: {
              layers: {
                foo: {
                  id: 'foo',
                  layout: { visibility: 'visible' },
                  filter: ['featureProperty', '==', 'bar', 'foo'],
                },
              },
            },
          },
          layers: {
            foo: {
              id: 'foo',
              source: 'features',
              type: 'circle',
              filter: ['featureType', 'Point'],
              paint: { 'circle-color': '#00FF00' },
              layout: { visibility: 'none' },
            },
          },
        },
      };
      deepFreeze(state);

      const layerConfig = getLayerConfig(state, 'bar', 'foo');

      expect(layerConfig).toEqual({
        id: 'foo',
        type: 'circle',
        source: 'features',
        layout: { visibility: 'visible' },
        paint: { 'circle-color': '#00FF00' },
        filter: ['featureProperty', '==', 'bar', 'foo'],
      });
    });
  });

  describe('isFeatureOnLayer()', () => {
    describe('layer.features', () => {
      it('returns true if the feature id is in the layers features prop', () => {
        const feature = { id: 'foo' };
        const layer = {
          id: 'layer1',
          features: ['foo'],
        };

        expect(isFeatureOnLayer(feature, layer)).toBe(true);
      });

      it('returns false if the feature id is not the layers features prop', () => {
        const feature = { id: 'foo' };
        const layer = {
          id: 'layer1',
          features: ['bar'],
        };

        expect(isFeatureOnLayer(feature, layer)).toBe(false);
      });

      it('returns false if there are no features or filter', () => {
        const feature = { id: 'foo' };
        const layer = {
          id: 'layer1',
        };

        expect(isFeatureOnLayer(feature, layer)).toBe(false);
      });
    });

    describe('layer.filter', () => {
      it('calls filter.featureProperty when featureProperty filter is found', () => {
        const feature = point([0, 0]);
        const layer = { id: 'layer', filter: [] };

        isFeatureOnLayer(feature, layer);

        expect(runLayerFilter).toHaveBeenCalledWith(feature, layer.filter);
      });

      it('doesnt call filter.featureProperty when featureProperty filter is not found', () => {
        const feature = point([0, 0]);
        const layer = { id: 'layer' };

        isFeatureOnLayer(feature, layer);

        expect(runLayerFilter).not.toHaveBeenCalled();
      });

      it('returns false if the layer doesnt exist', () => {
        expect(isFeatureOnLayer({ id: 'foo' }, undefined)).toBe(false);
      });
    });
  });

  describe('getGeneratedStyleForMap()', () => {
    it('returns generated style when map and style exist', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {
              generatedStyle: {
                foo: 'bar',
              },
            },
          },
        },
      };

      const style = getGeneratedStyleForMap(state, 'map1');

      expect(style).toEqual({ foo: 'bar' });
    });

    it('returns undefined when map does not exist', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: [],
          },
        },
      };

      const style = getGeneratedStyleForMap(state, 'map1');

      expect(style).toBeUndefined();
    });

    it('returns undefined when style doesnt exist on map', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {},
          },
        },
      };

      const style = getGeneratedStyleForMap(state, 'map1');

      expect(style).toBeUndefined();
    });

    it('returns undefined when mapId wasnt passed', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {},
          },
        },
      };

      const style = getGeneratedStyleForMap(state);

      expect(style).toBeUndefined();
    });
  });

  describe('getFlyToPositionForMap()', () => {
    it('returns center position when map and flyTo exist', () => {
      const state = {
        '@prisma/map': {
          map: {
            map1: {
              flyTo: {
                latitude: 36,
                longitude: -77,
              },
            },
          },
        },
      };

      const center = getFlyToPositionForMap(state, 'map1');

      expect(center).toEqual({ latitude: 36, longitude: -77 });
    });

    it('returns empty object when map does not exist', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: [],
          },
        },
      };

      const center = getFlyToPositionForMap(state, 'map1');

      expect(center).toEqual({});
    });

    it('returns empty object when flyTo doesnt exist on map', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {},
          },
        },
      };

      const center = getFlyToPositionForMap(state, 'map1');

      expect(center).toEqual({});
    });

    it('returns empty object when mapId wasnt passed', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {},
          },
        },
      };

      const center = getFlyToPositionForMap(state, 'map1');

      expect(center).toEqual({});
    });
  });

  describe('getMapModeForMap(mapId)', () => {
    it('returns mode when map exists', () => {
      const state = {
        '@prisma/map': {
          map: {
            map1: {
              mode: 'draw',
            },
          },
        },
      };

      const mode = getModeForMap(state, 'map1');

      expect(mode).toEqual('draw');
    });

    it('returns null when map does not exist', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: [],
          },
        },
      };

      const mode = getModeForMap(state, 'map1');

      expect(mode).toBeNull();
    });

    it('returns normal when mode doesnt exist on map', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {},
          },
        },
      };

      const mode = getModeForMap(state, 'map1');

      expect(mode).toEqual('normal');
    });

    it('returns null when mapId wasnt passed', () => {
      const state = {
        '@prisma/map': {
          map: {
            mapIds: ['map1'],
            map1: {},
          },
        },
      };

      const mode = getModeForMap(state);

      expect(mode).toBeNull();
    });
  });
});
