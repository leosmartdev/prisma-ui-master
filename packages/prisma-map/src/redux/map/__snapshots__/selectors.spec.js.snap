// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/map/selectors createStylesForAllMaps add features to all maps that match 1`] = `
Object {
  "bar": Object {
    "layers": Array [
      Object {
        "id": "layer2",
        "source": "generated:layer2",
      },
      Object {
        "id": "layer5",
        "source": "generated:layer5",
      },
    ],
    "mapId": "bar",
    "sources": Object {
      "generated:layer2": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer5": Object {
        "data": Object {
          "features": Array [
            Object {
              "geometry": Object {
                "coordinates": Array [
                  0,
                  1,
                ],
                "type": "Point",
              },
              "properties": Object {
                "layerId": "layer3",
              },
              "type": "Feature",
            },
          ],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "url": Object {
        "type": "geojson",
        "url": "http://foo.com",
      },
    },
  },
  "baz": Object {
    "layers": Array [
      Object {
        "id": "layer2",
        "source": "generated:layer2",
      },
      Object {
        "id": "layer4",
        "source": "generated:layer4",
      },
    ],
    "mapId": "baz",
    "sources": Object {
      "generated:layer2": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer4": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "url": Object {
        "type": "geojson",
        "url": "http://foo.com",
      },
    },
  },
  "foo": Object {
    "layers": Array [
      Object {
        "id": "layer1",
        "source": "url",
      },
      Object {
        "id": "layer2",
        "source": "generated:layer2",
      },
      Object {
        "id": "layer3",
        "source": "generated:layer3",
      },
      Object {
        "id": "layer4",
        "source": "generated:layer4",
      },
      Object {
        "id": "layer5",
        "source": "generated:layer5",
      },
    ],
    "mapId": "foo",
    "sources": Object {
      "generated:layer2": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer3": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer4": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer5": Object {
        "data": Object {
          "features": Array [
            Object {
              "geometry": Object {
                "coordinates": Array [
                  0,
                  1,
                ],
                "type": "Point",
              },
              "properties": Object {
                "layerId": "layer3",
              },
              "type": "Feature",
            },
          ],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "url": Object {
        "type": "geojson",
        "url": "http://foo.com",
      },
    },
  },
}
`;

exports[`redux/map/selectors createStylesForAllMaps adds features to proper layer sources 1`] = `
Object {
  "foo": Object {
    "layers": Array [
      Object {
        "id": "layer1",
        "source": "url",
      },
      Object {
        "id": "layer2",
        "source": "generated:layer2",
      },
    ],
    "mapId": "foo",
    "sources": Object {
      "generated:layer2": Object {
        "data": Object {
          "features": Array [
            Object {
              "geometry": Object {
                "coordinates": Array [
                  0,
                  1,
                ],
                "type": "Point",
              },
              "properties": Object {
                "bar": "baz",
              },
              "type": "Feature",
            },
            Object {
              "geometry": Object {
                "coordinates": Array [
                  1,
                  2,
                ],
                "type": "Point",
              },
              "properties": Object {
                "bar": "foo",
              },
              "type": "Feature",
            },
          ],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "url": Object {
        "type": "geojson",
        "url": "http://foo.com",
      },
    },
  },
}
`;

exports[`redux/map/selectors createStylesForAllMaps only adds a feature to the first layer that matches 1`] = `
Object {
  "foo": Object {
    "layers": Array [
      Object {
        "id": "layer1",
        "source": "url",
      },
      Object {
        "id": "layer2",
        "source": "generated:layer2",
      },
      Object {
        "id": "layer3",
        "source": "generated:layer3",
      },
      Object {
        "id": "layer4",
        "source": "generated:layer4",
      },
    ],
    "mapId": "foo",
    "sources": Object {
      "generated:layer2": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer3": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer4": Object {
        "data": Object {
          "features": Array [
            Object {
              "geometry": Object {
                "coordinates": Array [
                  0,
                  1,
                ],
                "type": "Point",
              },
              "properties": Object {
                "layerId": "layer3",
              },
              "type": "Feature",
            },
          ],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "url": Object {
        "type": "geojson",
        "url": "http://foo.com",
      },
    },
  },
}
`;

exports[`redux/map/selectors getStylesForLoadedMaps() adds all sources and layers listed in dynamicLayers 1`] = `
Object {
  "foo": Object {
    "layers": Array [
      Object {
        "id": "layer1",
        "source": "generated:layer1",
        "type": "circle",
      },
      Object {
        "id": "layer2",
        "source": "generated:layer2",
        "type": "fill",
      },
    ],
    "mapId": "foo",
    "sources": Object {
      "generated:layer1": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
      "generated:layer2": Object {
        "data": Object {
          "features": Array [],
          "type": "FeatureCollection",
        },
        "type": "geojson",
      },
    },
  },
}
`;

exports[`redux/map/selectors getStylesForLoadedMaps() only returns generated styles for loaded maps 1`] = `
Object {
  "bar": Object {
    "layers": Array [],
    "mapId": "bar",
    "sources": Object {},
  },
  "foo": Object {
    "layers": Array [],
    "mapId": "foo",
    "sources": Object {},
  },
}
`;
